### PR TITLE
FIX:  bug in 'reflect' and 'wrap' coordinate mapping

### PR DESCRIPTION
**Pull request by grlee77**
This pull request fixes subtle indexing errors in ``coord_map`` from ``_shared.interpolation.pyx``
for modes 'reflect' and 'mirror'.

I added a small python wrapper to coord_map so I could create a test.  The image example shown below was added to the documentation.

The following lists output for a set of 1D coordinates before and after the PR:

```python
wrap = [coord_map(4, n, 'W') for n in range(-6, 6)]
# before PR, wrap = [3, 1, 2, 3, 1, 2, 0, 1, 2, 3, 1, 2]
# after PR, wrap  = [2, 3, 0, 1, 2, 3, 0, 1, 2, 3, 0, 1]
```
Note that coordinate 0 was not wrapped!

```python
reflect = [coord_map(4, n, 'R') for n in range(-6, 6)]
# before PR, reflect = [0, 1, 2, 3, 2, 1, 0, 1, 2, 3, 2, 1]
# after PR, reflect  = [2, 3, 3, 2, 1, 0, 0, 1, 2, 3, 3, 2]
```
note that the edge coordinates 0 and 3 were not being duplicated in the mirrors

The difference is perhaps easier to see in images than the 1D example above. The original image extent is indicated by the region inside the yellow lines.

**Prior to the PR:**
buggy_edge_modesv2

**After this PR:**
fixed_edge_modesv2

### CODE REVIEW

**Comment by grlee77** [comment]
As an aside, is there any easy way during development to force cython to regenerate all .c files from the .pyx files?  Several .pyx files in the codebase include `interpolation.pxd` which contains inline functions.  Because the .pyx files themselves have not been changed, cython does not think it needs to regenerate the corresponding .c files.

For now, I just manually deleted the .c files and used cythonize from the command line to regenerate them.  Is there a simpler way?

**Comment by JDWarner** [comment]
Thanks for the PR and the very informative figure! That really makes it clear (and I like it as an example, too).

Regarding cleaning compiled files, on platforms with makefile support `make clean` will scour the package clean of compiled files.

**Comment by grlee77** [comment]
@JDWarner:
``make clean``removes the .so files, but this isn't enough in this case.  I needed to also remove the .c files generated by cython so they can be regenerated using the updated version of ``interpolation.pxd``.  Normally if you change a .pyx file cython will detect that and regenerate the .c file, but it does not appear to check whether any included .pxd files have changed as well.

**Comment by JDWarner** [comment]
I was just looking at the `Makefile` and you're right. I'll update that file to be more comprehensive.

In the meantime you can run the command behind `make clean` directly, with `"*.c"` and `".pyd"` added.:

``` bash
find . -name "*.so" -o -name "*.pyc" -o -name "*.pyx.md5" -o -name "*.c" -o -name "*.pyd" | xargs rm -f
```

**Comment by grlee77** [comment]
@JDWarner:  
Okay, but it is still a bit more complicated than the above.  We need to not remove any .c files that were not generated by cython.  I know there are at least a couple of these:
``restoration/unwrap_2d_ljmu.c``
``restoration/unwrap_3d_ljmu.c``

**Comment by grlee77** [comment]
Now that I think about it more, it could be argued that the original 'reflect' behaviour is correct if you consider the point of reflection to be the **center** of the edge pixels as opposed to their edges.  

The original 'wrap' behaviour does not seem correct to me, though.

**Comment by jni** [comment]
@grlee77 I would indeed consider `reflect`'s behaviour correct, because it matches NumPy's:

```python
In [139]: np.pad([0, 1, 2, 3], 2, mode='reflect')
Out[139]: array([2, 1, 0, 1, 2, 3, 2, 1])
```

**Comment by grlee77** [comment]
Thanks @jni:
It does seem the np.pad wrap behaviour matches what I have proposed here, though

```python
In[43]:    np.pad([0, 1, 2, 3], 4, mode='wrap')
Out[43]: array([0, 1, 2, 3, 0, 1, 2, 3, 0, 1, 2, 3])
```
What I implemented as reflect in this PR is equivalent to ``np.pad``'s "symmetric":
```python
In[44]:    np.pad([0, 1, 2, 3], 4, mode='symmetric')
Out[44]: array([3, 2, 1, 0, 0, 1, 2, 3, 3, 2, 1, 0])
```

As a further comparison, I looked at the behaviour implemented in the function ``map_coordinate`` from ``scipy/ndimage/src/ni_interpolation.c``.  That function also accepts non-integer coordinates, but from what I can tell, when integer inputs are used, mode 'reflect' matches what I have implemented in this pull request.  However, the ``map_coordinate`` function from ``ndimage`` also has a mode called 'mirror' which gives a result matching the behaviour of 'reflect' prior to my pull request!

Oddly, the 'wrap' behavior in ``map_coordinate`` seems to match the pre-PR wrap behaviour above!  Am I missing something here or is this a bug?  

I don't have a strong opinion on which reflect mode should be used (perhaps offer both as options?).  This all came up because I was looking into the potential impact of changing the default mode for ``transform.resize`` and ``transform.rescale`` to 'reflect' as discussed in #1514.

**Comment by grlee77** [comment]
see #1585 for the case that lead me to start looking into this.  (The images shown there look identical whether using current master or the code from this PR because the problem seems to be in the 'reflect' mode of ndi.map_coordinates itself)

**Comment by ahojnnes** [comment]
Thanks very much for this bug fix! This is extremely important.

I hope to have some time later today to look over this.

**Comment by grlee77** [comment]
As implemented in the most recent commits, I propose to rename the old pre-PR 'reflect' to 'mirror' for consistency with ``ndimage``.  The behaviour for 'reflect' will become what is proposed in this PR and would be consistent with ndimage (but consistent with mode 'symmetric' of ``np.pad`` & ``skimage.util.pad``).  

Several other functions in ``skimage`` already accept both 'mirror' and 'reflect' modes (via calls to ``ndimage`` functions) .  For example:
https://github.com/scikit-image/scikit-image/blob/master/skimage/feature/corner.py#L60
https://github.com/scikit-image/scikit-image/blob/master/skimage/transform/pyramids.py#L48
https://github.com/scikit-image/scikit-image/blob/master/skimage/filters/_gaussian.py#L28

The proposed 'reflect' behavior is also what is implemented in dask's array ghosting (used by ``skimage.util.apply_parallel``):
https://github.com/ContinuumIO/dask/blob/master/dask/array/ghost.py#L301-L333

However, this still leaves the question of which is the desired behaviour for mode 'wrap'.  The proposed behavior matches ``np.pad`` and ``dask`` and is more intuitive to me, but the previous behaviour matches ``ndimage``.

Here is a summary of the current behavior in this PR:
all_modes_v2

**Comment by grlee77** [comment]
For completeness, the following code demonstrates the behaviour in ``ndi.map_coordinates``.  Comparing to my previous comment above, only mode 'wrap' is different.

```python
import scipy.ndimage as ndi
from matplotlib import pyplot as plt

img = np.zeros((9, 9))
img[:8, :8] += 1
img[:4, :4] += 1
img[:2, :2] += 1
img[:1, :1] += 2
img[8, 8] = 4

rows, cols = img.shape[0], img.shape[1]
map_rows, map_cols = np.mgrid[-rows:2*rows, -cols:2*cols]
coord_map = np.array([map_rows, map_cols])
order=1, cval=0
fig, axes = plt.subplots(1, 5, figsize=(15, 5))
for n, mode in enumerate(['constant', 'nearest', 'wrap', 'reflect', 'mirror']):
    out = ndi.map_coordinates(img, coord_map, order=order,
                              mode=mode, cval=cval)
    axes[n].imshow(out, cmap=plt.cm.gray, interpolation='nearest')
    axes[n].plot([8.5, 8.5], [8.5, 17.5], 'y--', linewidth=0.5)
    axes[n].plot([17.5, 17.5], [8.5, 17.5], 'y--', linewidth=0.5)
    axes[n].plot([8.5, 17.5], [8.5, 8.5], 'y--', linewidth=0.5)
    axes[n].plot([8.5, 17.5], [17.5, 17.5], 'y--', linewidth=0.5)
    axes[n].set_axis_off()
    axes[n].set_aspect('equal')
    axes[n].set_title(mode)
plt.tight_layout()
```
**ndimage.map_coordinates result**:
all_modes_ndimage_v2

File: doc/examples/plot_edge_modes.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    axes[n].plot([17.5, 17.5], [8.5, 17.5], 'y--', linewidth=0.5)
+    axes[n].plot([8.5, 17.5], [8.5, 8.5], 'y--', linewidth=0.5)
+    axes[n].plot([8.5, 17.5], [17.5, 17.5], 'y--', linewidth=0.5)
+    axes[n].set_axis_off()
+    axes[n].set_aspect('equal')
+    axes[n].set_title(mode)
+
+plt.tight_layout()
+
+plt.show()
```
Comment (ahojnnes) [inline doc/examples/plot_edge_modes.py]: Add newline

File: skimage/_shared/_interpolation_test.pyx [hunk 2]
```diff
@@ -0,0 +1,31 @@
+from interpolation cimport coord_map as _coord_map
```
Comment (ahojnnes) [inline skimage/_shared/_interpolation_test.pyx]: Would you mind doing the following:

interpolation.pyx -> interpolation_cy.pyx
interpolation.py (wrapper for coord_map)
Make the _interpolation_test.pyx a unit test and move to _shared/tests/test_interpolation.py

Comment (ahojnnes) [inline skimage/_shared/_interpolation_test.pyx]: Ah, ok. I should have looked further below, you already created the unit test. So, forget the last line... :-)

File: skimage/_shared/tests/test_interpolation.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
+from skimage._shared._interpolation_test import coord_map
+from numpy.testing import assert_array_equal
+
+
+def test_coord_map():
+
```
Comment (ahojnnes) [inline skimage/_shared/tests/test_interpolation.py]: remove line

File: skimage/transform/tests/test_warps.py [hunk 4]
```diff
@@ -181,6 +181,18 @@ def test_resize3d_resize():
...
+
+def test_resize3d_2din_3dout():
+    # 3D output with 2D input
+    x = np.zeros((5, 5), dtype=np.double)
+    x[1, 1] = 1
+    resized = resize(x, (10, 10, 1), order=0)
+    ref = np.zeros((10, 10, 1))
+    ref[2:4, 2:4] = 1
+    assert_almost_equal(resized, ref)
+
```
Comment (ahojnnes) [inline skimage/transform/tests/test_warps.py]: Remove one line

**Comment by ahojnnes** [comment]
Just a few comments, but all looks good! Thanks for the fix.

Regarding the mismatch in terminology between skimage, np.pad and scipy.ndimage: Can we make the behavior consistent with np.pad and internally make it consistent with scipy.ndimage whenever we use it for 3D (replace the `modes`)?

**Comment by grlee77** [comment]
@ahojnnes

See if you are satisfied with the naming/organization of _shared/interpolation.pyx now. 

It certainly would be possibile to do the following transformation of ``np.pad`` mode names to the corresponding ``ndimage`` behaviour.
'reflect' -> 'mirror'
'symmetric' -> 'reflect'
However, there would still be the other places in the codebase that wrap ndimage functionality, using those naming conventions and not the ones from np.pad.  Would these all then need to be modified accordingly? I am not sure if this would help or make things more confusing for users.

Does anyone else have an opinion on the best approach to take here?

**Comment by grlee77** [comment]
I have also (re)raised the issue of the unexpected wrap behaviour in ``scipy.ndimage`` here:
https://github.com/scipy/scipy/issues/2640#issuecomment-120079980

**Comment by ahojnnes** [comment]
Thanks for your amazing work. Let's wait for an opinion from at least one other core developer on this before you spend any additional time.

**Comment by jni** [comment]
@grlee77 I don't think anyone in their right mind would think ndimage's wrap is correct. That's definitely a bug in my book.

I don't have a strong opinion about reflect/mirror vs symmetric/reflect, because both are in use by pretty major functions in this ecosystem. It's a shame that they diverged. Some thoughts to consider:

- `map_coordinates` is a closer function to `interpolate`, so it makes sense to mirror their API. (heh. =P)
- `np.pad` is a more "primal" function in the scipy universe, so it makes sense to use that convention.

I'm leaning towards the second consideration, because I think there might be more of a chance to get ndimage to match numpy than vice-versa, especially since they are on 0.15 (not 1.x yet), while numpy is on 1.9, ie, mature. So we could bring about a unification, rather than continuing the schism.

Whatever we end up deciding, thank you @grlee77 for this!

**Comment by jni** [comment]
I'm particularly impressed by your demo figures, they are super-clear, and make reviewing so much easier. It might be worth putting them in our docs!

**Comment by JDWarner** [comment]
My guess is these were independently developed. I can't take credit (or blame) for the `np.pad` API - I just made it fast ;) - but to cloud the picture even more:

* `np.pad` is in arguably the most core package of the ecosystem, but wasn't introduced until 1.7
* `ndi.map_coordinates` has existed for much longer.

I prefer `np.pad`'s conventions but that's likely just familiarity talking, so I should recuse my vote here.

**Comment by grlee77** [comment]
We need a decision on the names for the border modes to move this forward.  In summary, what was being called 'reflect' in the current codebase, matches the behavior of mode 'mirror' of ``ndimage`` (but does match 'reflect' of ``np.pad``).  No equivalent of ``ndimage`` 'reflect' (or ``np.pad`` 'symmetric') was previously provided.

In the current PR, I have made 'mirror' and 'reflect' match the ``ndimage`` behaviour.  I think this approach is the simplest since a number of other routines in ``skimage`` already take ``ndimage`` mode arguments.  However, it was suggested to perhaps use the border mode names from ``np.pad`` (and ``skimage.util.pad``) instead.  There is no clear consensus on this from the core devs who responded.  Do any of the other core devs want to chime in?  @stefanv maybe?

If we go with the ``np.pad`` names there would be 3 names that would require a name conversions whenever routines from ``ndimage`` are called :
- 'edge' -> 'nearest'
- 'reflect' -> 'mirror'
- 'symmetric' -> 'reflect'

np.pad modes below:
numpy_pad_modes

ndimage modes (with bug in wrap mode):
ndimage_modes

OpenCV has slightly different names as well:
opencv_border_modes

**Comment by ahojnnes** [comment]
@grlee77 Thanks again for putting this together. I earlier mentioned that we should replicate the np.pad behavior. @stefanv @JDWarner @jni Would you please give some feedback here? Thanks.

**Comment by JDWarner** [comment]
I also favor the np.pad conventions.

**Comment by ahojnnes** [comment]
I think, we can go ahead and wrap the np.pad naming scheme. Gregory, would
you be willing to translate the relevant wrap modes in the warp function?
Thanks!

**Comment by grlee77** [comment]
@ahojnnes: sure, I will update it soon

**Comment by grlee77** [comment]
@ahojnnes 
Please check over the naming changes.  I think I got everything and the tests are passing for me, but something could have been missed.

Travis failure on Python 2.6 seems to be due to #1629

There are a number of other functions not based on `warp` or `_shared.interpolation` that use the `ndimage` conventions instead.  I did not change those in this PR.  You can find these if you search skimage for the string 'mirror' which is a mode that exists in `ndimage` but is called 'reflect' in `numpy.pad`

If this looks good to you, I can do a quick rebase to rename the commits properly via BUG:, MAINT:, TST:, etc.

File: skimage/_shared/interpolation.pxd [hunk 5]
```diff
@@ -2,6 +2,20 @@
 #cython: boundscheck=False
 #cython: nonecheck=False
 #cython: wraparound=False
+"""
+Note: All edge modes implemented here follow the corresponding numpy.pad
```
Comment (ahojnnes) [inline skimage/_shared/interpolation.pxd]: Very nice.

File: skimage/transform/_warps_cy.pyx [hunk 6]
```diff
@@ -70,20 +70,29 @@ def _warp_fast(cnp.ndarray image, cnp.ndarray H, output_shape=None,
...
+        Points outside the boundaries of the input are filled according
+        to the given mode.  Modes match the behaviour of `numpy.pad`.
     cval : string, optional (default 0)
         Used in conjunction with mode 'C' (constant), the value
         outside the image boundaries.
 
+    Note
+    ----
+    Modes 'reflect' and 'symmetric' are similar, but differ in whether the edge
+    voxels are duplicated during the reflection.  As an example, if an array
```
Comment (ahojnnes) [inline skimage/transform/_warps_cy.pyx]: voxel -> pixel?

File: skimage/_shared/interpolation.pyx [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+
+def coord_map_py(Py_ssize_t dim, long coord, mode):
+    """ interpolation.coord_map python wrapper """
+    cdef char mode_c = ord(mode[0].upper())
+    return coord_map(dim, coord, mode_c)
+
+
+def extend_image(image, pad=10, mode='constant', cval=0):
+    """ Pad a 2D image by ``pad`` pixels on each side.
```
Comment (ahojnnes) [inline skimage/_shared/interpolation.pyx]: Minor: Remove space at beginning.

File: skimage/util/apply_parallel.py [hunk 8]
```diff
@@ -68,13 +68,18 @@ def apply_parallel(function, array, chunks=None, depth=0, mode=None,
...
+    mode : {'reflect', 'symmetric', 'periodic', 'wrap', 'nearest', 'edge'}, optional
+        type of external boundary padding.
     extra_arguments : tuple, optional
         Tuple of arguments to be passed to the function.
     extra_keywords : dictionary, optional
         Dictionary of keyword arguments to be passed to the function.
 
+    Notes
+    -----
+    Numpy edge modes `symmetric`, `wrap` and `edge` are converted to the
```
Comment (ahojnnes) [inline skimage/util/apply_parallel.py]: I guess this should be "scipy.ndimage edge modes ...."? I thought we replicate np.pad behavior?

**Comment by ahojnnes** [comment]
@grlee77 Left you some minor comments. Other than that it looks great!

File: skimage/transform/_warps_cy.pyx (hunk 6 above)
Comment (grlee77) [inline skimage/transform/_warps_cy.pyx]: yes, I will change it.  I work a lot with MRI where the use of "voxel" is common :)

File: skimage/util/apply_parallel.py (hunk 8 above)
Comment (grlee77) [inline skimage/util/apply_parallel.py]: This change is not directly related to the interpolation name changes in this PR, I just noticed it when checking all of these modes.  This function calls [dask](https://github.com/ContinuumIO/dask) which has yet another set of names than either `ndimage` or `numpy.pad`.  I just updated it so any `numpy.pad` names that have an equivalent in `dask` are converted properly.

Comment (ahojnnes) [inline skimage/util/apply_parallel.py]: Ah okay. But then you need to wrap those modes in warp as well, whenever we fall back to ndimage?

Comment (ahojnnes) [inline skimage/util/apply_parallel.py]: Forget what I said. You already do that. Sorry for the confusion.

File: skimage/transform/_geometric.py [hunk 9]
```diff
@@ -1294,7 +1296,7 @@ def warp(image, inverse_map=None, map_args={}, output_shape=None, order=1,
     >>> warped = warp(cube, coords)
 
     """
-
+    mode = _mode_deprecations(mode)
```
Comment (ahojnnes) [inline skimage/transform/_geometric.py]: Further below, we fall back to ndimage. The modes need to be wrapped there as well.

File: skimage/_shared/utils.py [hunk 10]
```diff
@@ -163,3 +163,14 @@ def assert_nD(array, ndim, arg_name='image'):
...
     if not array.ndim in ndim:
         raise ValueError(msg % (arg_name, '-or-'.join([str(n) for n in ndim])))
+
+
+def _mode_deprecations(mode):
+    """ to be used by functions to update deprecated mode names in
+    `skimage._shared.interpolation.pyx`."""
+    if mode.lower() == 'nearest':
+        warnings.warn(skimage_deprecation(
+            "Mode 'nearest' has been renamed 'edge'. Mode 'nearest' will be "
```
Comment (ahojnnes) [inline skimage/_shared/utils.py]: Please, add a new item to TODO.txt

**Comment by grlee77** [comment]
Hmmm... I somehow got one extra commit that is not mine included here now in the rebase.  probably should have left it alone!  Is there an easy way to fix this?

**Comment by grlee77** [comment]
okay just rebased off of current master and I think that may have fixed it.

File: skimage/transform/_geometric.py (hunk 9 above)
Comment (ahojnnes) [inline skimage/transform/_geometric.py]: Still not solved.

**Comment by stefanv** [comment]
I apologize for taking so long into this PR.  First, thanks so much for the effort you put into such clear explanations, Gregory.  Now, this issue comes a long way.  In fact, I should probably take the blame, because I re-enabled the "mirror" mode in ndimage which the original author, Peter Verveer, labeled with a warning comment.  Here Be Dragons?  Meh.

So, the problem, partially, is that SciPy has what I think is a problematic representations of pixels as "tiny little blocks".  Thinking about this off and on over a few years, I've come to expect values to represent data points in space.  I.e., the first element of a numpy image array represents a value at the coordinate `(0, 0)`.

This issue is important enough that I mentioned it in the [SciPy roadmap](https://github.com/rgommers/scipy/blob/roadmap/doc/ROADMAP.rst.txt).  Solving it will close no less than four (!) SciPy bugs.

I think the scikit-image pixel model is more sane than ndimage's, and that we should aim at fixing SciPy if we can (and the broken modes highlighted above, of course) since we fall back to it quite often.

Finally, I am delighted to see that my inattention to the issue really made no difference, since you all figure this and more out yourselves--kudos!

File: skimage/transform/_geometric.py (hunk 9 above)
Comment (grlee77) [inline skimage/transform/_geometric.py]: Good catch.  I will move `_to_ndimage_mode` into that file and import from there in warps.py

**Comment by grlee77** [comment]
@ahojnnes:  see if the additions to TODO.txt seems adequate (I think maybe I should have put it under 0.14 instead), then hopefully we are ready to merge!

**Comment by ahojnnes** [comment]
LGTM. Approval from my side. Waiting for a second voice from another core dev @JDWarner @stefanv @jni etc.

**Comment by JDWarner** [comment]
LGTM as well, @ahojnnes. Thanks for sticking with this, @grlee77!

**Comment by ahojnnes** [comment]
Thanks very much @grlee77 - this fix is very important!

Since you seem to be familiar with the interpolation code now, how about a new challenge: https://github.com/scikit-image/scikit-image/issues/1279 :-)

Anyway, thanks very much.

**Comment by stefanv** [comment]
@grlee77 @ahojnnes Can you please have a look at https://github.com/scipy/scipy/issues/5223 and confirm whether scikit-image is doing the right thing?  The question is whether:

```
tf.rescale(np.array([1,2,3,4.]), (1, 3), order=0)
```

is supposed to return

```
array([[ 1.,  1.,  1.,  2.,  2.,  2.,  3.,  3.,  3.,  4.,  4.,  4.]])
```

as it currently does?

**Comment by ahojnnes** [comment]
That's what it is supposed to do. skimage uses the convention that the pixel centers are at 0.5. Is that enough?

**Comment by stefanv** [comment]
@ahojnnes I thought we agreed to use the data point system?  The centers of pixels system seems a bit odd, and causes a lot of problems in, e.g., scipy.ndimage.

**Comment by ahojnnes** [comment]
I have a strong opinion of having the centers of pixels as (0.5, 0.5) - it
has always been like this in scikit-image and to my knowledge it is the
standard in any other image processing package. Otherwise you'll end up
with wierd effects for the first and last pixels in a row / column when
rescaling images.

**Comment by ahojnnes** [comment]
What's your issue with that convention?

**Comment by stefanv** [comment]
It assumes square pixels, for one.  And it doesn't generalize very well
when pixel spacing changes.

**Comment by ahojnnes** [comment]
I can see that, but I considers these as edge cases in which you have to
use ndimage.geometric_transform anyway. I would argue that 99% of the
images for which scikit-image is used have square pixels.

**Comment by stefanv** [comment]
It's fine, but we should document the convention in the user guide.
I'll file an issue.

**Comment by ahojnnes** [comment]
Agreed :-)
